core/modules/replica/IngestRequestMgr.cc

```diff
@@ -0,0 +1,313 @@
+
+        // Make the best effort to clean up the temporary files (if any) left after previous
+        // run of the unfinished requests. Requests that are eligible to be resumed will
+        // open new empty files as they will be being processed.
+        if (cleanupOnResume && !contrib.tmpFile.empty()) {
+            boost::system::error_code ec;
+            fs::remove(contrib.tmpFile, ec);
+            if (ec.value() != 0) {
+                LOGS(_log, LOG_LVL_WARN, context_ << "file removal failed for: '" << contrib.tmpFile
+                    << "', error: '" << ec.message() << "', ec: " + to_string(ec.value()));
+            }
+        }
+
+        // Note that an actual state of the in-progress contribution requests in which:
+        //
+        //   contrib.status == TransactionContribInfo::Status::IN_PROGRESS
+        //
+        // is determined by the series of the timestamps (from the oldest to the most recent):
+        //
+        //   contrib.createTime -- is guaranteed to be non-zero for any request recorded in
+        //     the database. At this stage the request was sitting in the input queue waiting
+        //     to be picked up by the next available processing thread. If the request was still
+        //     in the queue then contrib.startTime, contrib.readTime and contrib.loadTime are
+        //     guaranteed to be set to 0.
+        //
+        //   contrib.startTime -- is set to a non-zero value for contributions pulled from the input
+        //     queue by a processing thread and put into the "in-progress" queue. The timestamp
+        //     is set at a moment when the thread finishes creating a temporary file where
+        //     the preprocessed content of the corresponding input file gets stored. Right after
+        //     setting the timestamp the thread begins reading the input file. Timestamps
+        //     contrib.readTime and contrib.loadTime are guaranteed to be set to 0 while the input
+        //     file is still being read.
+        //
+        //   contrib.readTime -- is set to a non-zero value by a processing thread after finishing
+        //     reading/preprocessing the input file and writing its preprocessed content into
+        //     the temporary file. Right after that the thread begins loading the content of
+        //     the file into MySQL. While the loading is still in progress the timestamp
+        //     contrib.loadTime is guaranteed to be set to 0
+        //
+        //   contrib.loadTime -- is set to a non-zero value by a processing thread after finishing
+        //     uploading the content of teh temporary file into MySQL. At this point the request
+        //     is supposed to be completed and moved into the output queue.
+        //
+        // The auto-resume algorithm is evaluting a progress of requests from the newest
+        // timestamps back to the oldest ones in order to determine at what stage each
+        // request was before the restart.
```

Comment:
That sort of implies that your clock is monotonic, which is hard to guarantee in real setup. May be better to try to avoid this dependency on time ordering, or use a special monotonic "clock". (but do not use std::steady_clock).

Reply:
I do not actually see that you are comparing timestamps, so newest/oldest above does not mean ordered by time. You can ignore what I said above.